shared/src/lib.rs

```diff
@@ -75,22 +91,87 @@ impl From<SocketAddr> for PravegaNodeUri {
 impl PravegaNodeUri {
     pub fn to_socket_addr(&self) -> SocketAddr {
         // to_socket_addrs will resolve hostname to ip address
-        let mut addrs_vec: Vec<_> = self
-            .0
-            .to_socket_addrs()
-            .expect("Unable to resolve domain")
-            .collect();
-        addrs_vec.pop().expect("get the first SocketAddr")
+        match PravegaNodeUri::uri_parts_from_string(self.to_string()) {
+            Ok(uri_parts) => {
+                let mut addrs_vec: Vec<_> =
+                    format!("{}:{}", uri_parts.authority.unwrap(), uri_parts.port.unwrap())
+                        .to_socket_addrs()
+                        .expect("Unable to resolve domain")
+                        .collect();
+                addrs_vec.pop().expect("get the first SocketAddr")
+            }
+            Err(e) => panic!("{}", e),
+        }
     }
 
     pub fn domain_name(&self) -> String {
-        let parts: Vec<_> = self.0.split(':').collect();
-        parts[0].to_string()
+        match PravegaNodeUri::uri_parts_from_string(self.to_string()) {
+            Ok(uri_parts) => uri_parts.authority.expect("uri missing authority"),
+            Err(e) => panic!("{}", e),
+        }
     }
 
     pub fn port(&self) -> u16 {
-        let parts: Vec<_> = self.0.split(':').collect();
-        parts[1].parse::<u16>().expect("parse port to u16")
+        match PravegaNodeUri::uri_parts_from_string(self.to_string()) {
+            Ok(uri_parts) => uri_parts.port.expect("parse port to u16"),
+            Err(e) => panic!("{}", e),
+        }
+    }
+
+    /// Return Result of the uri scheme or empty string if no scheme was specified
+    pub fn scheme(&self) -> PravegaNodeUriParseResult<String> {
+        match PravegaNodeUri::uri_parts_from_string(self.to_string()) {
+            Ok(sa) => match sa.scheme {
+                Some(scheme) => Ok(scheme),
+                _ => Ok("".to_string()),
+            },
+            Err(e) => Err(e),
+        }
+    }
+
+    /// verifies the uri is well-formed (contains at least host:port, with optional scheme:// prefix)
+    ///
+    pub fn is_well_formed(uri: String) -> bool {
+        match PravegaNodeUri::uri_parts_from_string(uri) {
+            Ok(uri_parts) => uri_parts.port.is_some() && uri_parts.authority.is_some(),
+            Err(_) => false,
+        }
+    }
+
+    fn uri_parts_from_string(uri: String) -> PravegaNodeUriParseResult<PravegaNodeUriParts> {
+        let mut uri_parts: PravegaNodeUriParts = PravegaNodeUriParts::default();
+        let mut parts: Vec<_> = uri.split("://").collect();
+        if parts.len() > 2 || parts.is_empty() {
+            return Err(PravegaNodeUriParseError::ParseError {
+                error_msg: "malformed uri".into(),
```

Comment:
Here and below include the bad URI in the error message.

Reply:
are you saying the error message will include the uri, or that it should include the uri?

I don't think the uri is currently being included in the error message. 

Are you sure you want the uri to be included in the error message. What if it includes userinfo (future functionality?)

Reply:
I think it should be included to aid debugging. We don't have userinfo now, and this would only be returning the string that was passed in back to the code that provided it, so I don't see an issue.

Reply:
k, I will get back to the PR  this weekend, will add this.

Reply:
I have added uri to the error messages.